design/build-cache.md

```diff
@@ -0,0 +1,201 @@
+
+#### [Tentative] Regarding commits, branches and scoping
+
+Caching needs to take into account commits and branches when performing scoping.
+
+Scoping a cache to commit level and/or not considering branches will lead to issues with caching in the following instances:
+
+- Build and create cache when deploying/build for commit in dev branch. Same cache gets matched when building commit for production branch, however production actually dynamically generates different resources/artifacts.
+- Have a single commit that gets deployed from a single PR or branch with multiple configurations.
+
+Scoping should therefore take into account:
+
+1. Branch
+2. Commit
+3. Hash signature of content to be cached
+
+#### Job level
+
+This level of caching explicitly links a cache to certain jobs.
+
+Use when:
+
+1. Injecting the cached resources into every job is not desirable.
+
+#### Event level
+
+Event level scoping is scoped to a single pipeline execution event. Cached data is visible and accessible to all jobs within a single execution of the pipeline.
+
+Executions of a pipeline for the same commit are viewed as different executions.
+
+Use when:
+
+1. Generated resources are time dependant
+2. Generated resources are only valid for single execution
+3. Generated resources are meant for deployment (Please note that triggering deployment with a detached job will create a new event and, by extension, cache.)
+
+#### Pipeline level
+
+Pipeline level scoping is scoped to a pipeline as a whole. Cached data is visible and accessible to all instances of the pipeline.
+
+Note: Concurrency for simultaneous builds need proper attention.
+
+Use when:
+
+1. Build dependencies take a long time to resolve
+2. Generated resources are meant for deployment (Risk prone, not recommended unless deleting deployed artifact from cache after deployment)
+
+### (Proposal) Local and remote caching
+
+Pipeline and event level caches are stored on the build cache service node. They are copied to the build node on pipeline setup, and back to the build cache on pipeline completion.
+
+Once copied to the build node, they are persisted as a shared volume, mounted onto the appropriate path in the build container. 
+
+This proposal is currently blocked by the nature of the backend implementation for nodes. The k8s executor for example, selects a completely random node for every job in the pipeline. Caching locally is therefore not viable at the moment.
+
+### Build cache service
+
+The build cache service is essentially a specialized version of the store service. It utilizes the builtin Hapi cache system backed by Catbox. 
```

Comment:
Would a network cache service be too slow in some cases? For example npm cache essentially uses local disk to cache network I/O. But if the entire npm cache has to be loaded through network when the build starts, doesn't it defeat the purpose? I can see gains from using tar to reduce the # of network calls, or from talking to a service in the same network instead of a remote NPM register, but the gains are probably not as significant as using disk volumes, and could be offset by the added size of the files that need to be loaded. We probably need to do some speed testing before going with this choice.

On the other hand, for cases where things have to be re-compiled (for example builds using source code dependencies like cocoapods for iOS or REPO for Android), a network cache service totally makes sense.

Reply:
I can see your point of build dependency cache deviates from the purpose of SD. But build dependencies within the same event is fine IMO, since you generally want to lock down the dependencies in the same event. Also even for dependencies across different events, it should still be fine as long as people are using dependency management tools that can resolve dependency versions correctly (e.g. npm).

Reply:
Yeah, it's a simple purist statement. I foresee caching of build dependencies on a pipeline level. (For things like npm or sbt managed dependencies)